Removing dev/proc/sys mount restriction from DiskSpaceCollector

JScott opened the pull request:
We're trying to collect data about a tmpfs mount at `/dev` and under `/sys` and diskspace.py will skip right over those mount points. There's no comment as to why this is done, unlike the previous checks, and given [issue#262](https://github.com/python-diamond/Diamond/issues/262) I wouldn't be surprised if the original purpose has been lost. We aren't seeing any issue when we manually do the heavy lifting:

```
>>> def test(string):
...     stat = os.stat(string)
...     major = os.major(stat.st_dev)
...     minor = os.minor(stat.st_dev)
...     print stat
...     print major
...     print minor
...
>>> test('/dev')
posix.stat_result(st_mode=16877, st_ino=1025, st_dev=5L, st_nlink=14, st_uid=0, st_gid=0, st_size=4280, st_atime=1456860792, st_mtime=1460050361, st_ctime=1460050361)
0
5
>>> test('/run')
posix.stat_result(st_mode=16877, st_ino=1212, st_dev=16L, st_nlink=20, st_uid=0, st_gid=0, st_size=700, st_atime=1456868938, st_mtime=1460151861, st_ctime=1460151861)
0
16
```

This check should be removed if it doesn't have a purpose. If it does then there should at least be a comment or abstraction to explain why this restriction is in place.

To give some context, these are our tmpfs mount locations that would be great to monitor:

```
$ sudo cat /proc/mounts | grep tmpfs
udev /dev devtmpfs rw,relatime,size=24650316k,nr_inodes=6162579,mode=755 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=4932236k,mode=755 0 0
none /sys/fs/cgroup tmpfs rw,relatime,size=4k,mode=755 0 0
none /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
none /run/shm tmpfs rw,nosuid,nodev,relatime 0 0
none /run/user tmpfs rw,nosuid,nodev,noexec,relatime,size=102400k,mode=755 0 0
```

JScott commented:
Also, if there is a purpose to the restriction then perhaps it can be a default that can be overridden. Something with `exclude_filters`maybe.

JScott commented:
Has anyone looked at this? It's a small change with a big impact and it'd be really unfortunate if we had to maintain our own fork because of 4 legacy lines.

shortdudey123 commented:
Looks like the if statement is original to the collector.  I personally don't see any reason to have it in there.  It does keep you from having tmpfs in there which most people don't care about (but some due).

Can you rebase on master to fix the merge conflict?  Also can you update the testing to include mount points that you are now allowing?

JScott commented:
Any tips you can give on writing the test? I'm going to spend a bit of time digging into [testdiskspace.py](https://github.com/python-diamond/Diamond/blob/master/src/collectors/diskspace/test/testdiskspace.py) today to see if I can grok the existing tests but guidance would be appreciated.

shortdudey123 commented:
The fixtures file `src/collectors/diskspace/test/fixtures/proc_mounts` contains sample outputs of `/proc/mounts`.  This is then loaded and the values in that file are verified to make sure the collector picked up the correct values.  I would add another method similar to the `test_should_work_with_tmpfs` one and use `/dev`, `/proc`, and `/sys` lines to verify the functionally of your changes.  The output you expect goes in the `metrics` dict.

JScott commented:
Failed on `pip install -r .travis.requirements.txt` which I haven't touched, is there a way to retest?

shortdudey123 commented:
@JScott fasted way i do it is with `git commit --amend` and force push.  The only ones who can restart the job are the people part of the python-diamond github org

JScott commented:
Any ideas what's going on here? It can't find cython during the Python 2.6 requirements install on pyutmp:

```
Collecting pyutmp (from -r .travis.requirements.txt (line 16))
  Using cached pyutmp-0.2.1.zip
    sh: 1: cython: not found
    Failed to build "pyutmp_linux.c" from "pyutmp_linux.pyx"
    "pyutmp_linux.pyx" is newer than "pyutmp_linux.c". Rebuilding "pyutmp_linux.c".
    + cython pyutmp_linux.pyx
    Complete output from command python setup.py egg_info:
    sh: 1: cython: not found
    
    Failed to build "pyutmp_linux.c" from "pyutmp_linux.pyx"
    
    "pyutmp_linux.pyx" is newer than "pyutmp_linux.c". Rebuilding "pyutmp_linux.c".
    
    + cython pyutmp_linux.pyx
    
    ----------------------------------------
    Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-vU5wJc/pyutmp
```

It failed twice now, hours apart, so it's more than a fluke. It was working fine when I merged to master (0a2d2da) but before I wrote tests and abstracted a few lines from that test file.

shortdudey123 commented:
its been off and on for a while
adding cython the the req file probably will solve that

shortdudey123 commented:
LGTM :)

@jaingaurav can you review?

jaingaurav commented:
@JScott: Looks good, I can merge. Could you merge all the commits into a single commit and update. You can keep the Cython requirement change as a separate commit (no need to make it into a separate PR).

JScott commented:
@jaingaurav can I get some advice on squashing? Making 2 commits out of the last 3 is easy enough but  2dbadc4 is behind the wall of commits from when I had to merge downstream. It looks like it'd be easier for me to destroy and remake my fork and repeat all the work again than to squash the earlier commit.

shortdudey123 commented:
@JScott 
`git rebase -i HEAD~3` then force push

JScott commented:
@shortdudey123 if I go back the last 5 commits I did then I get a list like this due to my previous merge into master:

```
pick 2a204fd Add SSL and host header support for nginx collector
pick 2dbadc4 Removing dev/proc/sys mount restriction from DiskSpaceCollector
pick b59df4e [jolokia] Add timeout when making HTTP requests.
pick 08e8d96 [jolokia] Request just the JMX domain names when making a list request.
pick 9e4900c [jolokia] Don't recompile the regexps every time
pick 24f3aca [jolokia] The list of JMX domains can be set in the config file.
pick 05b41cf [jolokia] Add support for 'canonicalNaming' jolokia parameter.
pick 41bf9bd Fix broken tests.
pick 42dbd99 [jolokia] Made more methods private. Fixed typos.
pick 7b31edb [jolokia] Added option to specify the path of jolokia service.
pick a3bd5fa Add `_normalized` metrics for loadaverage
pick 23a2b84 Add memory.used_percent for RedisCollector
pick 48fe1e5 Gitter badge and travis integration
pick c026288 Update docs
pick ad043f5 Add cluster_health.status metrics for ElasticSearch
pick e832399 Add CentOS 7 build and dev hosts to vagrant
pick f1714ec fix bad docker name join
pick 722a1a7 Fix fluentd collector
pick 5ea5f76 Add maxCollectionSize to support legacy versions of Jolokia
pick 8a47fc7 remove README.md from MANIFEST.in
pick 98dd405 Add ntp collector
pick 0945ec5 Fix in setup.py for oracle linux
pick 13ef0d1 Support tmpfs in diskspace collector
pick 578e79e Adding a test for diskspace in system directories
pick 9f79ce4 Adding Cython to the requirements
```

I've merged the last 3 down to 2, which is the easiest change to make. If you want me to merge in the other commit (2dbadc4) then I'll either need a way to ignore commits that aren't mine in an interactive rebase or refork. I can go up to `~50` and still not get back to that commit because it was a month before I got an initial response and had to manually handle merging.

JScott commented:
That recent failure seems to be due to Cython again. It was passing before I force pushed it so I'll have to look into that again I suppose.

jaingaurav commented:
@JScott: So I'm no git expert but if you do an interactive rebase can't you squish 2dbadc4 & 0a2d2da? I think there is an unnecessary merge commit there.

Thanks for doing this. As soon as the commits are fixed I'll pull in right away. Main thing I want, is to try and keep tests passing on individual commits.

JScott commented:
@jaingaurav Not that I know although I'm probably as much of an expert as you. I think if I had branched this in my fork and rebased instead of merging onto master then I could just rebase locally onto master and have all my commits together. As it stands though, the history on my side is 2dbadc4, the month of history merged by 0a2d2da, and then the rest of the stuff that I've already squashed.

The only test that's failed right now is 9f79ce4 which is the most recent, easily tweakable. If passing commits is your concern then I shouldn't need to squash the old commit. However, if you want it all together then I can remake the changes with another PR since I didn't really change much once the dust settled.

shortdudey123 commented:
If you want, i can send you an email and we can do a hangout and i can walk you though it :)
Its actually very useful to know how to do rebasing and squashing!

JScott commented:
@shortdudey123 I know how to rebase and squash but if you think you have insight for this particular squashing a messy, merged history problem then I'd love to pair on it. Do you have any time between 1-5 PST today? My GitHub profile has my email.

JScott commented:
@jaingaurav Squashed!

shortdudey123 commented:
Yay 🎉

JScott commented:
@jaingaurav can I get this merged yet?

m-messiah commented:
Does anybody know, why it is not merged yet?

shortdudey123 commented:
@m-messiah probably because it has not been reviewed yet :)
